elastic-buffer-sim/src/Bittide/Topology.hs

```diff
@@ -25,77 +22,282 @@ module Bittide.Topology
+plotMesh ::
+  forall n.
+  (KnownNat n, 1 <= n, n <= 20) =>
+  Mesh n ->
+  Int ->
+  Int ->
+  IO ()
+plotMesh api@Mesh{..} m k = do
+  offs <- genOffs
+  uncurry (matplotWrite name)
+    $ unzip
+    $ map plotDats
+    $ simulateMesh defBittideClockConfig api m k offs
  where
-  (0, n) = A.bounds g
-  g = complete 6
+  plotDats =
+      bimap
+        (uncurry plot . P.unzip)
+        (foldPlots . P.fmap (uncurry plot . P.unzip) . L.transpose)
+    . P.unzip
+    . fmap (\(x,y,z) -> ((x,y), (x,) <$> z))
+
+-- | Given a mesh with \(n\) vertices, creates a function which takes
+-- a bound @m@ for the number of samples, a bound @k@ for the sampling
+-- period, a vector of \(n\) offsets (divergence from spec) and a
+-- vertex @i@ selected for simulation.
+--
+-- NOTE: All clocks are implicitly synchronized to the same clock
+-- domain at this point, which is ok as long as the dynamic clock
+-- generator is used. Otherwise the domains have to be differentiated
+-- at the type level, which is not straightforward to archive for fully
+-- connected topologies.
+simulateMesh ::
+  forall n m dom.
+  ( Clash.KnownDomain dom
+  , KnownNat n
+  , KnownNat m
+  , 1 <= n
+  , 1 <= m
+  , n + m <= 32
+  , 1 + n <= 32
+  ) =>
+  ClockControlConfig dom m ->
+  Mesh n ->
+  Int ->
+  Int ->
+  Vec n Offset ->
+  Vec n [(Period, Period, [DataCount m])]
+simulateMesh ccc Mesh{..} m k !offsets =
+    pointwise
+  $ P.take m
+  $ takeEveryN k
+  $ absTimes available
+  $ bundle
+  $ zipWith (curry bundle) clkSignals (bundle <$> ebs)
+ where
+  -- elastic buffers
+  !ebs = imap ebv clocks
+  ebv x = flip imap clocks . eb x
+  eb x xClk y yClk
+    | available x y = elasticBuffer Error xClk yClk
+    | otherwise     = pure 0
+  -- clocks
+  !clocks = clock <$> offsets <*> clockControls
+  clock offset =
+    tunableClockGen
+      (cccSettlePeriod ccc)
+      offset
+      (cccStepSize ccc)
+      Clash.resetGen
+  -- clock controls
+  !clockControls = clockControl <$> clocks <*> masks <*> ebs
+  clockControl clk =
+    callistoClockControl
+      clk
+      Clash.resetGen
+      Clash.enableGen
+      ccc
+    . pure
+  -- clock signals
+  !clkSignals = extractPeriods <$> clocks
+  -- available link mask vectors
+  !masks = nVec (v2bv . nVec . avail)
+  avail x y = if available x y then high else low
+  nVec = flip map indicesI
+
+absTimes ::
+  (KnownNat n, NFData a) =>
+  (Index n -> Index n -> Bool) ->
+  Signal dom (Vec n (Period, Vec n a)) ->
+  [Vec n (Period, Period, [a])]
+absTimes available = go $ replicate SNat (Clash.Femtoseconds 0)
+ where
+  go !ts (v Clash.:- vs) =
+    force (izipWith (\i t (p, es) -> (t, p, filterAvailable i es)) ts v)
+      : go (force $ zipWith addFs ts $ map fst v) vs
+  filterAvailable i = catMaybes . toList . imap (asMaybe . available i)
+  asMaybe = \case
+    True  -> Just
+    False -> const Nothing
+
+-- | Push infinity to the vector element level
+pointwise :: (KnownNat n, NFData a) => [Vec n a] -> Vec n [a]
```

Comment:
Isn't this just:

```suggestion
transpose :: (KnownNat n, NFData a) => [Vec n a] -> Vec n [a]
```

Reply:
Why do we need the `NFData`?

Reply:
We don't. Its a leftover from my "plugging the space leak" efforts. I'll take it out.

Reply:
I don't like to change the name to `transpose`, since it name-clashes with the vector `transpose` and may confuse people. My alternative was `elementwise`, but it looks strange as a function name.

Reply:
Fair point. `transposeV` perhaps? Otherwise feel free to ignore.